### PR TITLE
Test changing IP address of 2 nodes in a cluster out of 3 & misc cleanups

### PR DESCRIPTION
**Pull request by kostja**

### CODE REVIEW

File: test/pylib/scylla_cluster.py [hunk 1]
```diff
@@ -519,7 +533,7 @@ def __str__(self):
         host_id = getattr(self, 'host_id', 'undefined id')
         return f"ScyllaServer({self.server_id}, {self.ip_addr}, {host_id})"
 
-    def _write_config_file(self) -> None:
+    def write_config_file(self) -> None:
```
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:536]: I recall asking you why did you remove the `_` prefix for a private method on the other PR.
You said that it was by mistake and reverted the change.
Now you bring it back. Why?

File: test/pylib/scylla_cluster.py [hunk 2]
```diff
@@ -838,6 +852,24 @@ def setLogger(self, logger: logging.LoggerAdapter):
         for srv in self.servers.values():
             srv.setLogger(self.logger)
 
+    async def change_ip(self, server_id: ServerNum) -> ActionReturn:
+        """Lease a new IP address and update conf/scylla.yaml with it. The
+        original IP is released at the end of the test to avoid an
+        immediate recycle within the same cluster."""
```
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:860]: We should probably test immediate recycle at some point too.
But we shouldn't hope for this to work until we start identifying nodes by UUID everywhere and rejecting any messages we receive whose destination UUID doesn't match ours.
Just noting for the future.

File: test/pylib/scylla_cluster.py [hunk 3]
```diff
@@ -838,6 +852,24 @@ def setLogger(self, logger: logging.LoggerAdapter):
...
+        if server_id not in self.servers:
+            return ScyllaCluster.ActionReturn(success=False, msg=f"Server {server_id} unknown")
+        server = self.servers[server_id]
+        if server.is_running:
+            msg = f"Server {server_id} is running: stop it first and then change its ip"
+            return ScyllaCluster.ActionReturn(success=False, msg=msg)
```
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:866]: It's awkward that we turn this into error code `404` (in `_server_change_ip`).

Since the server must be stopped - it is a precondition, and not satisfying it would be a bug in the test itself - I think we can just `assert` inside `ScyllaServer.change_ip` and don't have to check it here. If the assertion fires, it will be turned into error code `500`.

(We could say the same for the check above... but that's a larger refactor)

Anyway, the precondition that the server identified by `server_id` must be stopped should also be stated in the docstring for `ScyllaCluster.change_ip`.

File: test/pylib/driver_util.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (C) 2022-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+import logging
+from test.pylib.manager_client import ManagerClient
+from cassandra.cluster import Session                   # type: ignore # pylint: disable=no-name-in-module
+
+
+async def reconnect_driver(manager: ManagerClient) -> Session:
```
Comment (kbr-scylla) [inline test/pylib/driver_util.py:13]: BTW. `scylladb/python-driver#170` is already fixed and included in the dbuild docker image.
Driver should always reconnect if IPs don't change, and using `wait_for_cql` and `wait_for_cql_and_get_hosts` should suffice to ensure that the driver reconnects to a given host.

I just didn't get to removing this `reconnect_driver` workaround from Raft upgrade tests.

But the driver won't handle changing the IPs of all nodes. Then it may simply be impossible for the driver to reconnect - it's trying to use old IPs after all, in which case we need to connect it manually using a new set of IPs. But that's not a bug, it's expected.

In that case we should probably update the docstring, explaining when we may need to reconnect the driver.

Comment (kbr-scylla) [reply test/pylib/driver_util.py:13]: Also maybe it's better to keep those utils in one place? We have test/pylib/utils.py and some of the functions there are already related to the driver.

File: test/pylib/scylla_cluster.py (hunk 1 above)
Comment (kostja) [reply test/pylib/scylla_cluster.py:536]: Because I lost my edit during a godzillion of rebases!-)

File: test/pylib/scylla_cluster.py (hunk 3 above)
Comment (kostja) [reply test/pylib/scylla_cluster.py:866]: HTTP 500 is harder to debug. It would be nice to assert on manager_client, but it doesn't contain the server state right now which is sad.

Comment (kostja) [reply test/pylib/scylla_cluster.py:866]: I added the comment.

File: test/pylib/driver_util.py (hunk 4 above)
Comment (kostja) [reply test/pylib/driver_util.py:13]: 2. I dropped the patch from this series. I will take a look if it is still necessary for test_change_all

Comment (kbr-scylla) [reply test/pylib/driver_util.py]: > I will take a look if it is still necessary for test_change_all

It probably is, the driver doesn't know the new IPs of the cluster. If at least one node does not change its IP the driver should fetch IPs of others from there, but not if every node changes their IPs.

My point is, in this scenario this is no longer a workaround for scylladb/python-driver#170.

**Comment by kbr-scylla** [comment]
Qd
